frontends/smvparser.y

```diff
@@ -511,33 +665,62 @@ basic_expr: constant {
-                  smt::Term res = enc.solver_->make_term(smt::BVSmod, a->getTerm(), b->getTerm());
-                  $$ = new SMVnode(res);
+                  SMVnode::BVtype bvs_a = a->getBVType();
+                  SMVnode::BVtype bvs_b = b->getBVType();
+                  if (bvs_a == bvs_b == SMVnode::Unsigned){
+                    smt::Term res = enc.solver_->make_term(smt::BVUrem, a->getTerm(), b->getTerm());
+                    $$ = new SMVnode(res,SMVnode::Unsigned);
+                  } else if (bvs_a == bvs_b == SMVnode::Signed){
+                    smt::Term res = enc.solver_->make_term(smt::BVSmod, a->getTerm(), b->getTerm());
+                    $$ = new SMVnode(res,SMVnode::Signed);
+                  } else{
+                    throw CosaException ("Unsigned/Signed unmatch");
+                  } 
               }
             }
             | basic_expr OP_SHIFTR basic_expr{
               SMVnode *a = $1;
               SMVnode *b = $3;
               smt::Term res = enc.solver_->make_term(smt::BVLshr, a->getTerm(), b->getTerm());
-              $$ = new SMVnode(res);
+              $$ = new SMVnode(res,a->getBVType());
             }
             | basic_expr OP_SHIFTL basic_expr{
               SMVnode *a = $1;
               SMVnode *b = $3;
               smt::Term res = enc.solver_->make_term(smt::BVShl, a->getTerm(), b->getTerm());
-              $$ = new SMVnode(res);
+              $$ = new SMVnode(res,a->getBVType());
             }
             | basic_expr sizev {
               throw CosaException("No word1");
             }
             | basic_expr OP_CON basic_expr  {
               SMVnode *a = $1;
               SMVnode *b = $3;
-              smt::Term res = enc.solver_->make_term(smt::Concat, a->getTerm(), b->getTerm());
-              $$ = new SMVnode(res);
+              SMVnode::BVtype bvs_a = a->getBVType();
+              SMVnode::BVtype bvs_b = b->getBVType();
+              if(bvs_a != bvs_b){
+                throw CosaException("Unsigned/Signed unmatch");
+              } else{
+                smt::Term res = enc.solver_->make_term(smt::Concat, a->getTerm(), b->getTerm());
+                $$ = new SMVnode(res,SMVnode::Unsigned);
+              }               
+            }
+            | basic_expr "[" integer_val "]" {
+                throw CosaException("No index Subscript");
             }
             | basic_expr "[" basic_expr ":" basic_expr "]"{
-              throw CosaException("No word1");
+                SMVnode *a = $1;
+                SMVnode *b = $3;
+                SMVnode *c = $5;
+                SMVnode::BVtype bvs_a = a->getBVType();
+                SMVnode::BVtype bvs_b = b->getBVType();
+                SMVnode::BVtype bvs_c = c->getBVType();
+                if(bvs_a == SMVnode::BVnot || bvs_b != SMVnode::Integer || bvs_c != SMVnode::Integer){
+                  smt::Term res = enc.solver_->make_term(smt::Extract, a->getTerm(), b->getTerm(),c->getTerm());
```

Comment:
Also, this isn't the right way to apply `Extract` actually. `Extract` is an indexed operator, so you'd do it like this:
```
enc.solver_->make_term(smt::Op(smt::Extract, 
                                                          <b as a C++ integer>, 
                                                          <c as a C++ integer>), 
                                           a->getTerm());
```

Would it be possible to add a test for this? I think what might help clarify some of the confusion on this part is how Extract can be used e.g. the `[high_bit:low_bit]` syntax in nuXmv. The `high_bit` and `low_bit` cannot be variables, they have to be integer values, e.g. `[2:0]` is fine but not `[a:b]`. So you shouldn't be using `b` and `c` as terms, they should be C++ integers.